libraries/lib-graphics/graphics/Rect.h

```diff
@@ -0,0 +1,267 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  Rect.h
+
+  Dmitry Vedenko
+
+**********************************************************************/
+#pragma once
+
+#include "Point.h"
+#include "Size.h"
+
+#include <algorithm>
+
+namespace graphics
+{
+//! Rectangle that is defined by the top left corner and the size.
+template <typename DataType>
+struct RectType final
+{
+   //! Origin of the rectangle.
+   PointType<DataType> origin;
+   //! Size of the rectangle.
+   SizeType<DataType> size;
+
+   //! Return true if the rectangle is not empty.
+   bool IsValid() const noexcept
+   {
+      return size.width > 0 && size.height > 0;
+   }
+
+   //! Return true if the rectangle contains the point specified.
+   bool Contains(PointType<DataType> pt) const noexcept
+   {
+      return pt.x >= origin.x && pt.y >= origin.y &&
+             (pt.x < (origin.x + size.width)) &&
+             (pt.y < (origin.y + size.height));
+   }
+};
+
+//! Calculates the intersection of two rectangles.
+template <typename DataType>
+RectType<DataType>
+Intersect(RectType<DataType> firstRect, RectType<DataType> secondRect) noexcept
+{
+   if (firstRect.size.IsZero() || secondRect.size.IsZero())
+      return {};
+
+   const PointType<DataType> p11 = firstRect.origin;
+   const PointType<DataType> p12 = { firstRect.origin.x + firstRect.size.width,
+                                     firstRect.origin.y +
+                                        firstRect.size.height };
+   const PointType<DataType> p21 = secondRect.origin;
+   const PointType<DataType> p22 = {
+      secondRect.origin.x + secondRect.size.width,
+      secondRect.origin.y + secondRect.size.height
+   };
+
+   if (p12.x < p21.x || p11.x > p22.x || p11.y > p22.y || p12.y < p21.y)
```

Comment:
I think `>=` and `<=` are correct.  `p12` and `p22` are excluded points of the rectangles.